aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cu

```diff
@@ -826,6 +826,55 @@ Tensor& orgqr_helper_cusolver(Tensor& result, const Tensor& tau) {
   return result;
 }
 
+void apply_lu_cusolver_looped(Tensor& self, Tensor& pivots, Tensor& infos, bool get_pivots) {
+  auto infos_data = infos.data_ptr<int>();
+
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(self.scalar_type(), "lu_cusolver", [&]{
+    auto self_data = self.data_ptr<scalar_t>();
+    auto self_stride = matrixStride(self);
+    int batch_size = cuda_int_cast(batchCount(self), "batch size");
+    auto handle = at::cuda::getCurrentCUDASolverDnHandle();
+
+    int m = cuda_int_cast(self.size(-2), "m");
+    int n = cuda_int_cast(self.size(-1), "n");
+    int lda = std::max<int>(1, m);
+    for (int batch = 0; batch < batch_size; ++batch) {
+      if (get_pivots) {
+        auto pivots_data = pivots.data_ptr<int>();
+        auto pivots_matrix_stride = pivots.size(-1);
+        at::cuda::solver::getrf<scalar_t>(
+          handle, m, n,
+          self_data + batch * self_stride,
+          lda,
+          pivots_data + batch * pivots_matrix_stride,
+          infos_data + batch
+        );
+      }
+      else {
+        at::cuda::solver::getrf<scalar_t>(
+          handle, m, n,
+          self_data + batch * self_stride,
+          lda,
+          NULL,
+          infos_data  + batch
+        );
+      }
+    }
+  });
+
+  // Necessary because cuSOLVER uses nan for outputs that correspond to 0 in MAGMA.
```

Comment:
This should be done only for `get_pivots=false`, right?

Reply:
No not necessarily. See this comment: https://pytorch.slack.com/archives/CFZ7V4DNZ/p1619401824137300?thread_ts=1619363616.136700&cid=CFZ7V4DNZ

Reply:
Running the code from the comment but changing `NULL` to `dipiv` in `cusolverDnDgetrf` call shows that the pivoted version of getrf works fine and doesn't have nan in the output. So this block of code for replacing nan with a finite number should be run only for `get_pivots=false`.


<details>
  <summary>Code for checking the pivoted version</summary>
  
compile with `nvcc -lsucolver`
  
```c
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include <cuda_runtime.h>
#include <cusolverDn.h>
#include <cusolver_common.h>

int main() {

  int major=-1,minor=-1,patch=-1;
  cusolverGetProperty(MAJOR_VERSION, &major);
  cusolverGetProperty(MINOR_VERSION, &minor);
  cusolverGetProperty(PATCH_LEVEL, &patch);
  printf("CUSOLVER Version (Major,Minor,PatchLevel): %d.%d.%d\n", major,minor,patch);

  cusolverDnHandle_t handle=NULL;
  cudaStream_t stream=NULL;
  cusolverStatus_t status = CUSOLVER_STATUS_SUCCESS;
  int N = 3;
  double * cA = (double*)malloc(N * N * sizeof(double));
  double * dA;
  for (int i = 0; i < N * N; ++i) {
    cA[i] = 1;
  }
  cudaMalloc((void**)&dA, N * N * sizeof(double) );
  cudaMemcpy(dA, cA, N * N * sizeof(double), cudaMemcpyHostToDevice);

  assert(cusolverDnCreate(&handle) == CUSOLVER_STATUS_SUCCESS);
  assert(cudaStreamCreateWithFlags(&stream, cudaStreamNonBlocking) ==
         cudaSuccess);
  assert(cusolverDnSetStream(handle, stream) == CUSOLVER_STATUS_SUCCESS);
  int lwork;
  double *dwork;
  int *dipiv, *dinfo, *cinfo;
  cinfo = (int*)malloc(sizeof(int) * 1);
  cudaMalloc((void**)&dinfo, sizeof(int));
  cudaMalloc((void**)&dipiv, N * sizeof(int));
  status = cusolverDnDgetrf_bufferSize(handle,
                                       N,
                                       N,
                                       dA,
                                       N,
                                       &lwork);
  assert(CUSOLVER_STATUS_SUCCESS == status);
  cudaMalloc((void**)&dwork, sizeof(double) * lwork);
  cusolverDnDgetrf(handle,
                   N,
                   N,
                   dA,
                   N,
                   dwork,
                   dipiv,
                   dinfo);
  cudaMemcpy(cA, dA, N * N * sizeof(double), cudaMemcpyDeviceToHost);
  cudaMemcpy(cinfo, dinfo, sizeof(int), cudaMemcpyDeviceToHost);
  printf("dinfo: %d\n", *cinfo);
  for (int i = 0; i < N; ++i) {
    for (int j = 0; j < N; ++j) {
      printf("%lf ", cA[i + j * N]);
    }
    printf("\n");
  }

}
```
  
</details>

Reply:
Very sorry for that I wonder how I missed that.